### PR TITLE
wip: added test for user stake, user withdraw stake, reserve onlyAvat…

### PR DESCRIPTION
**Pull request by ordd**
…ar modifier methods without start / end

# Description
- Added test for user stake and withdraw stake.
- Added schemes for set the market maker, block interval and contribution address in the reserve contract
- Added tests for `onlyAvatar` functions in the reserve contract 

About # (link your issue here)
#125 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

# Checklist:
- [ ] PR title matches follow: (Feature|Bug|Chore) Task Name
- [ ] My code follows the style guidelines of this project
- [ ] I have followed all the instructions described in the initial task (check Definitions of Done)
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added reference to a related issue in the repository
- [ ] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request ( for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [ ] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

File: stakingModel/test/GoodFundManager.e2e.test.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    await goodFundManager.transferInterest(simpleStaking.address);
+    const gdPriceAfter = await marketMaker.currentPrice(cDAI.address);
+    expect(
+      Math.floor(gdPriceAfter.toNumber() / 100).toString()
+    ).to.be.equal(Math.floor(gdPriceBefore.toNumber() / 100).toString());
+  });
+
+  it("should create be able to stake dai", async () => {
+    await next_interval();
+    const stakingMock = await SimpleDAIStakingMock.new(
```
Comment (sirpy) [inline stakingModel/test/GoodFundManager.e2e.test.js]: you are not supposed to deploy ANY contracts, this is e2e tests, you test the contracts deployed in the migration scripts

**Review by sirpy — CHANGES_REQUESTED** [review]

File: contracts/identity/Identity.sol [hunk 2]
```diff
@@ -223,7 +223,8 @@ contract Identity is IdentityAdminRole, SchemeGuard, Pausable {
      * @param account the address to add
      */
     function _addWhitelisted(address account) internal {
-        whitelist.add(account);
+        if (!whitelist.has(account))
```
Comment (sirpy) [inline contracts/identity/Identity.sol]: this is very wrong. remove it. if user is whitelisted he can't be added again
same for removewhitelisted

File: stakingModel/test/GoodFundManager.e2e.test.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      from: staker
+    });
+    await simpleStaking
+    .stakeDAI(web3.utils.toWei("100", "ether"), {
+      from: staker
+    })
+    .catch(console.log);
+    await cDAI.exchangeRateCurrent();
+    let rate = await cDAI.exchangeRateStored();
+    expect(rate.toString()).to.be.equal("10201010101010101010101010101");
```
Comment (sirpy) [inline stakingModel/test/GoodFundManager.e2e.test.js]: 1. you are testing the functionality of the cDAI mock here. we dont test mocks. you test what you expect to happen from our contracts.
2. hard coded number without explanation

File: stakingModel/test/GoodFundManager.test.js [hunk 4]
```diff
@@ -183,6 +184,31 @@ contract(
...
+          from: staker
+        })
+        .catch(console.log);
+        await cDAI.exchangeRateCurrent();
+        let stakingGDBalanceBefore = await goodDollar.balanceOf(stakingMock.address);
+        await goodFundManager.transferInterest(stakingMock.address);
+        let stakingGDBalanceAfter = await goodDollar.balanceOf(stakingMock.address);
+        expect(
+          stakingGDBalanceAfter.sub(stakingGDBalanceBefore).toString()
+        ).to.be.equal("190329"); // interest that the staking contract recieved
```
Comment (sirpy) [inline stakingModel/test/GoodFundManager.test.js]: yes but why 190329, show the calculation in comment

File: stakingModel/test/GoodReserve.e2e.test.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    const cDAIBalanceReserveAfter = await cDAI.balanceOf(goodReserve.address);
+    expect((cDAIBalanceReserveAfter - cDAIBalanceReserveBefore).toString()).to.be.equal(
+      amount.toString()
+    );
+    expect((reserveBalanceAfter - reserveBalanceBefore).toString()).to.be.equal(
+      amount.toString()
+    );
+  });
+
+  it("should be able to buy gd with cDAI and the price should be the same", async () => {
```
Comment (sirpy) [inline stakingModel/test/GoodReserve.e2e.test.js]: WHAT? how can price be the same after selling/buying?

File: stakingModel/test/GoodReserve.e2e.test.js [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    let amount = 1e4;
+    let reserveToken = await marketMaker.reserveTokens(cDAI.address);
+    const cDAIBalanceBefore = await cDAI.balanceOf(founder);
+    const cDAIBalanceReserveBefore = await cDAI.balanceOf(goodReserve.address);
+    await goodDollar.approve(goodReserve.address, amount);
+    await goodReserve.sell(cDAI.address, amount, 0);
+    reserveToken = await marketMaker.reserveTokens(cDAI.address);
+    const cDAIBalanceAfter = await cDAI.balanceOf(founder);
+    const cDAIBalanceReserveAfter = await cDAI.balanceOf(goodReserve.address);
+    expect((cDAIBalanceAfter - cDAIBalanceBefore).toString()).to.be.equal("800000");
```
Comment (sirpy) [inline stakingModel/test/GoodReserve.e2e.test.js]: why 80000 show calculation

File: stakingModel/test/SimpleStaking.e2e.test.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    const addresses = deployment[NETWORK];
+    const DAI_ADDRESS = addresses["DAI"];
+    const cDAI_ADDRESS = addresses["cDAI"];
+    const STAKING_ADDRESS = addresses["DAIStaking"];
+    dai = await DAIMock.at(DAI_ADDRESS);
+    cDAI = await cDAIMock.at(cDAI_ADDRESS);
+    simpleStaking = await SimpleDAIStaking.at(STAKING_ADDRESS);
+  });
+
+  it("should mint new dai", async () => {
```
Comment (sirpy) [inline stakingModel/test/SimpleStaking.e2e.test.js]: we dont test mocks only our contracts

File: stakingModel/test/SimpleStaking.e2e.test.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+  it("should redeem cdai", async () => {
+    let cdaiBalance = await cDAI.balanceOf(staker);
+    await cDAI.redeem(cdaiBalance.toString(), { from: staker });
+    let balance = await dai.balanceOf(staker);
+    expect(balance.toString()).to.be.equal(web3.utils.toWei("100", "ether"));
+    dai.transfer(dai.address, balance.toString(), { from: staker });
+  });
+
+  it("should be able to withdraw stake by staker", async () => {
```
Comment (sirpy) [inline stakingModel/test/SimpleStaking.e2e.test.js:50]: 3 tests of mocks and one actuall test of our contract

File: stakingModel/test/UBIScheme.e2e.test.js [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    absoluteVote = await AbsoluteVote.at(dao_addresses.AbsoluteVote);
+    deploy_settings = await fse.readFile("../migrations/deploy-settings.json", "utf8");
+    // schemes
+    setReserve = await FundManagerSetReserve.new(avatarAddress, goodFundManager.address, goodReserve.address);
+    addMinter = await AddMinter.new(avatarAddress, goodReserve.address);
+  });
+  
+  it("should start the pool", async () => {
+    let isActive = await firstClaimPool.isActive();
+    if (!isActive)
```
Comment (sirpy) [inline stakingModel/test/UBIScheme.e2e.test.js]: this is e2e you test the state you expect. either it should be inactive on deploy or active

File: stakingModel/test/UBIScheme.e2e.test.js [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+      await firstClaimPool.start();
+    isActive = await firstClaimPool.isActive();
+    expect(isActive).to.be.true;
+  });
+
+  it("should be able to set the reserve as a minter", async () => {
+    let isMinter = await goodDollar.isMinter(goodReserve.address);
+    if (!isMinter) {
```
Comment (sirpy) [inline stakingModel/test/UBIScheme.e2e.test.js]: this is e2e you test the state you expect. either it should be minter on deploy or not

File: stakingModel/test/UBIScheme.e2e.test.js [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+    await dai.mint(web3.utils.toWei("1000", "ether"));
+    dai.approve(cDAI.address, web3.utils.toWei("1000", "ether"));
+    await cDAI.mint(web3.utils.toWei("1000", "ether"));
+    await cDAI.approve(goodReserve.address, amount);
+    await goodReserve.buy(cDAI.address, amount, 0);
+    let gdbalance = await goodDollar.balanceOf(founder);
+    await goodDollar.transfer(firstClaimPool.address, (Math.floor(gdbalance.toNumber() / 2)).toString());
+    await goodDollar.transfer(ubi.address, (Math.floor(gdbalance.toNumber() / 2)).toString());
+    expect((await goodDollar.balanceOf(ubi.address)).toNumber()).to.be.gt(0);
+    expect((await goodDollar.balanceOf(firstClaimPool.address)).toNumber()).to.be.gt(0);
```
Comment (sirpy) [inline stakingModel/test/UBIScheme.e2e.test.js]: what does testing transfer to ubi and firstclaim has to do with this test?
you want to test reserve can mint. the test finished when the buyer got his GD
you can test GD total cap has increased

File: stakingModel/test/UBIScheme.e2e.test.js [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+  });
+
+  it("should award a new user with the award amount on first time execute claim", async () => {
+    await increaseDays();
+    let isWhitelisted = await identity.isWhitelisted(claimer);
+    if (!isWhitelisted)
```
Comment (sirpy) [inline stakingModel/test/UBIScheme.e2e.test.js]: this is e2e you test the state you expect.

File: stakingModel/test/UBIScheme.e2e.test.js [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+    await dai.mint(web3.utils.toWei("1000", "ether"));
+    dai.approve(cDAI.address, web3.utils.toWei("1000", "ether"));
+    await cDAI.mint(web3.utils.toWei("1000", "ether"));
+    await cDAI.approve(goodReserve.address, amount);
+    await goodReserve.buy(cDAI.address, amount, 0);
+    let gdbalance = await goodDollar.balanceOf(founder);
+    await goodDollar.transfer(firstClaimPool.address, (Math.floor(gdbalance.toNumber() / 2)).toString());
+    await goodDollar.transfer(ubi.address, (Math.floor(gdbalance.toNumber() / 2)).toString());
+    let isWhitelisted = await identity.isWhitelisted(claimer);
+    if (!isWhitelisted)
```
Comment (sirpy) [inline stakingModel/test/UBIScheme.e2e.test.js]: same

**Review by sirpy — CHANGES_REQUESTED** [review]

File: stakingModel/test/GoodFundManager.e2e.test.js [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+  });
+
+  it("should collect the interest and transfer it to the reserve and recieves minted gd back to the staking contract", async () => {
+    await next_interval();
+    await cDAI.exchangeRateCurrent();
+    let recipientBefore = await goodDollar.balanceOf(ubiBridgeRecipient);
+    await goodFundManager.transferInterest(simpleStaking.address);
+    let recipientAfter = await goodDollar.balanceOf(ubiBridgeRecipient);
+    let stakingGDBalance = await goodDollar.balanceOf(simpleStaking.address);
+    expect(stakingGDBalance.toString()).to.be.equal("0"); //100% of interest is donated, so nothing is returned to staking
```
Comment (sirpy) [inline stakingModel/test/GoodFundManager.e2e.test.js:96]: need test for when avgDonatedInterest is not 100%, maybe not e2e, but in unit tests with SimpleStakingMock

**Review by sirpy — CHANGES_REQUESTED** [review]

File: stakingModel/test/GoodFundManager.test.js [hunk 15]
```diff
@@ -183,6 +184,33 @@ contract(
...
+      await dai.mint(staker, web3.utils.toWei("100", "ether"));
+      await dai.approve(stakingMock.address, web3.utils.toWei("100", "ether"), {
+        from: staker
+      });
+      await stakingMock
+        .stakeDAI(web3.utils.toWei("100", "ether"), {
+          from: staker
+        })
+        .catch(console.log);
+        await cDAI.exchangeRateCurrent();
```
Comment (sirpy) [inline stakingModel/test/GoodFundManager.test.js:204]:
```suggestion
        //this increases the exchangerate in our mock, so we can simulate interest
        await cDAI.exchangeRateCurrent();
```

File: stakingModel/test/UBIScheme.e2e.test.js (hunk 10 above)
Comment (sirpy) [reply stakingModel/test/UBIScheme.e2e.test.js]: not fixed yet

**Review by sirpy — CHANGES_REQUESTED** [review]
go over all coments

File: stakingModel/test/GoodFundManager.e2e.test.js (hunk 1 above)
Comment (ordd) [reply stakingModel/test/GoodFundManager.e2e.test.js]: the test moved to GoodFundManager.test.js
deadb071dc8af2d0955c9eafcdb8c31b3fd7556e

File: contracts/identity/Identity.sol (hunk 2 above)
Comment (ordd) [reply contracts/identity/Identity.sol]: it was a draft and not a final version, just for testing. it was replaced by using `authenticate`.
db43a7633912fed7aa89a7ff580fbe5238d650b5

File: stakingModel/test/GoodFundManager.e2e.test.js (hunk 3 above)
Comment (ordd) [reply stakingModel/test/GoodFundManager.e2e.test.js]: it's an irrelevant test, removed from the suggested version
db43a7633912fed7aa89a7ff580fbe5238d650b5

File: stakingModel/test/GoodFundManager.test.js (hunk 4 above)
Comment (ordd) [reply stakingModel/test/GoodFundManager.test.js]: db43a7633912fed7aa89a7ff580fbe5238d650b5

File: stakingModel/test/GoodReserve.e2e.test.js (hunk 5 above)
Comment (ordd) [reply stakingModel/test/GoodReserve.e2e.test.js]: it was a test for the precision and not for the price.
changed at db43a7633912fed7aa89a7ff580fbe5238d650b5

File: stakingModel/test/GoodReserve.e2e.test.js (hunk 6 above)
Comment (ordd) [reply stakingModel/test/GoodReserve.e2e.test.js]: added
db43a7633912fed7aa89a7ff580fbe5238d650b5

File: stakingModel/test/SimpleStaking.e2e.test.js (hunk 7 above)
Comment (ordd) [reply stakingModel/test/SimpleStaking.e2e.test.js]: db43a7633912fed7aa89a7ff580fbe5238d650b5

File: stakingModel/test/SimpleStaking.e2e.test.js (hunk 8 above)
Comment (ordd) [reply stakingModel/test/SimpleStaking.e2e.test.js:50]: db43a7633912fed7aa89a7ff580fbe5238d650b5

File: stakingModel/test/UBIScheme.e2e.test.js (hunk 10 above)
Comment (ordd) [reply stakingModel/test/UBIScheme.e2e.test.js]: db43a7633912fed7aa89a7ff580fbe5238d650b5

File: stakingModel/test/UBIScheme.e2e.test.js (hunk 11 above)
Comment (ordd) [reply stakingModel/test/UBIScheme.e2e.test.js]: firstclaim contract should has funds.
db43a7633912fed7aa89a7ff580fbe5238d650b5

File: stakingModel/test/UBIScheme.e2e.test.js (hunk 12 above)
Comment (ordd) [reply stakingModel/test/UBIScheme.e2e.test.js]: as mentioned - temporary, replaced with authenticate
db43a7633912fed7aa89a7ff580fbe5238d650b5

File: stakingModel/test/UBIScheme.e2e.test.js (hunk 13 above)
Comment (ordd) [reply stakingModel/test/UBIScheme.e2e.test.js]: temporary
db43a7633912fed7aa89a7ff580fbe5238d650b5

File: stakingModel/test/GoodFundManager.e2e.test.js (hunk 14 above)
Comment (ordd) [reply stakingModel/test/GoodFundManager.e2e.test.js:96]: already exists (80%)

File: stakingModel/test/UBIScheme.e2e.test.js (hunk 9 above)
Comment (ordd) [reply stakingModel/test/UBIScheme.e2e.test.js]: db43a7633912fed7aa89a7ff580fbe5238d650b5

**Pull request merged**